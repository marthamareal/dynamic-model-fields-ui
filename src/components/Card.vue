<template>
  <div class="col-md-4 mb-5" id="card">
    <div :class="ghostLoad">
      <b-spinner class="loader" variant="success"></b-spinner>
    </div>
    <div :class="contentClass">
      <div class="card-body">
        <h2 class="card-title">{{riskType.name}}</h2>
        <p class="card-text">{{riskType.description}}</p>
      </div>
      <div class="card-footer">
        <img class="edit" @click="onEdit" src="../assets/edit.png" v-b-tooltip.hover title="edit"/>
        <img class="delete" @click="onDelete(riskType.id)" src="../assets/delete.png" v-b-tooltip.hover title="delete"/>

        <b-button v-b-modal= "id" variant="primary">APPLY</b-button>
          <b-modal :id="id" hide-footer :title="riskType.name" >
            <Input :field="field" v-for="field in fields" :key="field.indexOf" :modalId="id"/>
            <b-button class="mt-3 btn-primary" block @click="handleSubmit">SUBMIT</b-button>
          </b-modal>
      </div>
    </div>
  </div>
</template>

<script>
import Input from './Input'

export default {
  name: 'Card',
  components: {
    Input
  },
  props: {
    riskType: Object,
    deleteRiskType: Function,
    handleResponse: Function
  },
  data (props) {
    console.log(this.isLoading)
    return {
      id: `modal-${props.riskType.id}`,
      fields: props.riskType.fields,
      deleteType: props.deleteRiskType,
      contentClass: 'card h-100',
      ghostLoad: 'card h-100 no-gh'
    }
  },
  methods: {
    handleSubmit () {
      this.handleResponse('Sorry', 'This feature is not available.', 'info')
    },
    onDelete (riskTypeId) {
      this.deleteType(riskTypeId)
      this.contentClass = 'card h-100 gh-content'
      this.ghostLoad = 'card h-100 ghost-load'
    },
    onEdit (e) {
      e.preventDefault()
      this.handleResponse('Sorry', 'This feature is not available.', 'info')
    }
  }
}
</script>

<style>
.gh-content {
  position: relative;
  z-index: -1;
  top: -310px;
}
.no-gh {
  display: none;
}
.ghost-load {
  position: relative;
  background-color: #e9ecefc7;
}
.loader {
  position: relative;
  top: 50%;
  left: 50%;
}
.modal-dialog {
  padding-top: 60px;
  min-width: 80px !important;
}
.btn-primary {
  color: #e8eaee;
  font-weight: 500;
  background-color: #35495e!important;
  border: 1px solid #7d7f827d;
  border-radius: 2px;
  float: right;
}
.btn-primary:hover {
  background-color: #35495ecb;
  border: 1px solid #35495e;
  color: #ecedef;
}
.card-title {
  border-bottom: 1px solid #dfdfdf;
  font-size: 25px;
  color: #35495e;
  min-height: 65px;
}
.card:hover {
  box-shadow: 0em 0em 0.5em #acacac8e;
  transition: box-shadow 0.6s;
}
.card-text {
  height: 95px;
  overflow-y: auto;
}
.card-footer {
  display: inline;
}
.card-footer img{
  padding: 6px;
  height: 30px;
  width: 30px;
  display: none;
}
.card:hover .card-footer img {
display: inline!important;
}

.card-footer img:hoover {
  box-shadow: 0em 0em 0.5em #3b3b3b7d;
  transition: box-shadow 0.6s;
}
</style>
