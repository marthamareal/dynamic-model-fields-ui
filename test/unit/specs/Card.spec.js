import Vue from 'vue'
import Card from '@/components/Card'
import { mount } from '../../../node_modules/@vue/test-utils'

import BSpinner from 'bootstrap-vue/es/components/spinner/spinner'
import BPagination from 'bootstrap-vue/es/components/pagination/pagination'
import BButton from 'bootstrap-vue/es/components/button/button'
import BModal from 'bootstrap-vue/es/components/modal/modal'
import vBModal from 'bootstrap-vue/es/directives/modal/modal'

Vue.directive('b-modal', vBModal)
Vue.component('b-modal', BModal)
Vue.component('b-button', BButton)
Vue.component('b-pagination', BPagination)
Vue.component('b-spinner', BSpinner)

jest.mock('axios', () => ({
    post: (options) => {
        const response = { 
            data: {
                "name": "test user",
                "Age": 11,
            }
        }
        return new Promise((resolve, reject) => resolve(response))
    }
  }))
  
describe('Navbar.vue', () => {
  const wrapper = mount(Card,{
      propsData: {   riskType: {
        "id": "10d98576-091f-4fb2-ad14-c40822748b51",
        "name": "Professional liability insurance",
        "description": "Professional liability insurance, also known",
        fileds: [
            {
            "id": "4b9fbf08-dd06-4320-b409-dc469c6fcb32",
            "name": "date",
            "field_type": "date",
            "options": null
            },
            {
            "id": "483a996e-e603-4437-b689-99193383a7e8",
            "name": "name",
            "field_type": "text",
            "options": null
            }
        ]
    },
    onResponse: jest.fn(),
    deleteRiskType: jest.fn(),
    }
  }
)

    it('should render as expected', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('should submit form', () => {
        const spy = jest.spyOn(wrapper.vm, 'onResponse')
        const applyButton = wrapper.find("button")
        applyButton.trigger('click')
        wrapper.find(".btn-block").trigger('click')
        expect(spy).toHaveBeenCalled()
    })

    it('should delete riskType', () => {
        const spy = jest.spyOn(wrapper.vm, 'deleteRiskType')
        const deleteIcon = wrapper.findAll("img").at(1)
        deleteIcon.trigger('click')
        expect(spy).toHaveBeenCalled()
    })

    it('should edit riskType', () => {
        const spy = jest.spyOn(wrapper.vm, 'onResponse')
        const editIcon = wrapper.findAll("img").at(0)
        editIcon.trigger('click', { preventDefault: jest.fn()})
        expect(spy).toHaveBeenCalled()
    })
})
