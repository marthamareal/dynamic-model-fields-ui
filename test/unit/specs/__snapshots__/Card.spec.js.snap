// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar.vue should render as expected 1`] = `
<div id="card" class="col-md-4 mb-5">
    <div class="card h-100 no-gh"><span aria-hidden="true" class="loader spinner-border text-success"><!----></span></div>
    <div class="card h-100">
        <div class="card-body">
            <h2 class="card-title">Professional liability insurance</h2>
            <p class="card-text">Professional liability insurance, also known</p>
        </div>
        <div class="card-footer"><img src="../assets/edit.png" title="" class="edit" data-original-title="edit"> <img src="../assets/delete.png" title="" class="delete" data-original-title="delete"> <button type="button" class="btn btn-primary">APPLY</button>
            <div>
                <div id="modal-10d98576-091f-4fb2-ad14-c40822748b51___BV_modal_outer_" style="position: absolute; z-index: 1040;">
                    <div id="modal-10d98576-091f-4fb2-ad14-c40822748b51" role="dialog" tabindex="-1" aria-hidden="true" class="modal fade" style="display: none;">
                        <div class="modal-dialog modal-md">
                            <div role="document" id="modal-10d98576-091f-4fb2-ad14-c40822748b51___BV_modal_content_" aria-labelledby="modal-10d98576-091f-4fb2-ad14-c40822748b51___BV_modal_header_" aria-describedby="modal-10d98576-091f-4fb2-ad14-c40822748b51___BV_modal_body_" class="modal-content">
                                <header id="modal-10d98576-091f-4fb2-ad14-c40822748b51___BV_modal_header_" class="modal-header">
                                    <h5 class="modal-title">Professional liability insurance</h5><button type="button" aria-label="Close" class="close">Ã—</button>
                                </header>
                                <div id="modal-10d98576-091f-4fb2-ad14-c40822748b51___BV_modal_body_" class="modal-body"> <button type="button" class="btn mt-3 btn-primary btn-secondary btn-block">SUBMIT</button></div>
                                <!---->
                            </div>
                        </div>
                    </div>
                    <!---->
                    <!---->
                </div>
            </div>
        </div>
    </div>
</div>
`;
